Ntasks per inst xxx

jedwards4b opened the pull request:
Add a derived variable NTASKS_PER_INST_COMP where COMP is the component name (ATM, LND, etc) 

Test suite:  scripts_regression_tests.py  
Test baseline: 
Test namelist changes: 
Test status: bit for bit

Fixes #1897 
User interface changes?:  N

Update gh-pages html (Y/N)?: N

Code review:

billsacks commented:
@jedwards4b thanks a lot for taking this on.

One minor issue is that a basic xmlquery gives extra output (note the NINST print):

```
$ ./xmlquery NTASKS_PER_INST_GLC
NINST = 1
	NTASKS_PER_INST_GLC: 4
```

However, a bigger set of issues is that it leaves out some functionality that is typically provided for other xml variables:

1. Doing a partial query doesn't work (`xmlquery -p NTASKS_PER`, for example)

2. Listing NTASKS_PER_INST for all components doesn't work (`xmlquery NTASKS_PER_INST`)

3. Trying to get documentation of this variable with something like `./xmlquery --full NTASKS_PER_INST_GLC` gives an error:

```
$ ./xmlquery --full NTASKS_PER_INST_GLC
NINST = 1

Results in group none
Traceback (most recent call last):
  File "./xmlquery", line 365, in <module>
    _main_func()
  File "./xmlquery", line 352, in _main_func
    if results[group][var]['desc'][0] is not None:
IndexError: list index out of range
```

4. Auto-generated web documentation will not be created for these variables

My impression is that solving these problems would require making these "true" xml variables, rather than variables that are created on demand. This could follow the example of other derived variables, such as TOTALPES.

If others ( @gold2718 @mvertens @dabail10 @alperaltuntas @bertinia  for example) feel that these limitations are okay, then I can live with this. But I'm reluctant to sign off on this without approval from others.

jedwards4b commented:
I don't like the idea of making this a true xml variable because that implies that the user can change it (for example the user can change TOTALPES but it has no effect) and because I think that we should not introduce new xml variable(s) as a solution to every situation.

billsacks commented:
> I think that we should not introduce new xml variable(s) as a solution to every situation.

I agree with this in a general, philosophical sense. But in this case, part of the issue was having this be something that users can query. In that case, it seems useful to have some documentation on this, so that they know it can be queried.

But if others weigh in and say that there's no need for documentation of this new set of variables, then I'll defer to that opinion.

billsacks commented on src/drivers/mct/cime_config/config_component.xml:
```diff
@@ -1913,6 +1913,24 @@
+    </values>
+    <group>mach_pes</group>
+    <file>env_mach_pes.xml</file>
+    <desc>number of tasks per instance for each component (readonly)</desc>
```
I'd like to see a bit more explicit documentation on how this is determined, as well as consistency with other options that say, "DO NOT EDIT". What about changing the description to:

```xml
<desc>Number of tasks per instance for each component. DO NOT EDIT: Set automatically by case.setup based on NTASKS, NINST and MULTI_DRIVER</desc>
```

jedwards4b commented:
Updated based on review - note that the comment in config_headers.xml about variables
MAX_TASKS_PER_NODE, MAX_MPITASKS_PER_NODE was incorrect and I have removed it.

billsacks commented:
Okay, this looks good now - thanks.

@jedwards4b did you run scripts_regression_tests on something close to the latest version (9beb46e or later), or only on the initial version?

jedwards4b commented:
@billsacks Just completed, all systems go.
